src/tmpfiles.c

```diff
@@ -279,182 +319,223 @@ static void tmpfiles(char *line)
-		mkparent(path, 0755);
-		rc = blkdev(path, mode ?: 0644, major, minor);
-		break;
-	case 'c':
-		rc = parse_mm(arg, &major, &minor);
-		if (rc)
-			break;
-		if (!strc) {
-			if (type[1] != '+')
-				break;
-			erase(path);
-		}
-		mkparent(path, 0755);
-		rc = chardev(path, mode ?: 0644, major, minor);
-		break;
-	case 'C':
-		if (!arg) {
-			paste(buf, sizeof(buf), "/usr/share/factory", path);
-			arg = buf;
+		case 'D':
+			if (fisdir(path)) {
+				nftw(path, do_delete, 20, FTW_DEPTH | FTW_PHYS);
+			}
+			break;
+		case 'e':
+		case 'f':
+		case 'F':
+		case 'l': /* Finit extension, like 'L' but only if target exists */
+		case 'L':
+		case 'p':
+			break;
+		case 'r':
+			rc = glob_do(path, erase);
+			if (rc && errno == ENOENT)
+				rc = 0;
+			break;
+		case 'R':
+			rc = glob_do(path, rmrf);
+			break;
+		case 'w':
+			break;
+		case 'X':
+		case 'x':
+			dbg("Unsupported x/X command, ignoring %s, no support for clean at runtime.", path);
+			break;
+		case 'Z':
+		case 'z':
+			break;
+		default:
+			errx(1, "Unsupported tmpfiles command '%s'", type);
+			return;
 		}
-		if (!strc) {
-			struct dirent **namelist;
-			int num;
+	}
 
-			if (!S_ISDIR(st.st_mode))
+	// file & directory creation logic
+	if (c_flag) {
+		switch (type[0]) {
+		case 'b':
+			rc = parse_mm(arg, &major, &minor);
+			if (rc)
 				break;
-			num = scandir(path, &namelist, NULL, NULL);
-			free(namelist);
-			if (num >= 3)
-				break; /* not empty */
-		}
-		if (fisdir(arg) && !fisslashdir(arg)) {
-			size_t len = strlen(arg) + 2;
-
-			dst = malloc(len);
-			if (!dst)
+			if (!strc) {
+				if (type[1] != '+')
+					break;
+				erase(path);
+			}
+			mkparent(path, 0755);
+			rc = blkdev(path, mode ?: 0644, major, minor);
+			break;
+		case 'c':
+			rc = parse_mm(arg, &major, &minor);
+			if (rc)
 				break;
-			snprintf(dst, len, "%s/", arg);
-			arg = dst;
-		}
-		mkparent(path, 0755);
-		rc = rsync(arg, path, LITE_FOPT_KEEP_MTIME, NULL);
-		if (rc && errno == ENOENT)
-			rc = 0;
-		break;
-	case 'd':
-	case 'D':
-		mkparent(path, 0755);
-		rc = mksubsys(path, mode ?: 0755, user, group);
-		break;
-	case 'e':
-		if (glob(path, GLOB_NOESCAPE, NULL, &gl))
-			break;
-
-		for (size_t i = 0; i < gl.gl_pathc; i++)
-			rc += mksubsys(gl.gl_pathv[i], mode ?: 0755, user, group);
-		break;
-	case 'f':
-	case 'F':
-		mkparent(path, 0755);
-		if (type[1] == '+' || type[0] == 'F') {
-			/* f+/F will create or truncate the file */
+			if (!strc) {
+				if (type[1] != '+')
+					break;
+				erase(path);
+			}
+			mkparent(path, 0755);
+			rc = chardev(path, mode ?: 0644, major, minor);
+			break;
+		case 'C':
 			if (!arg) {
-				rc = create(path, mode ?: 0644, user, group);
+				paste(buf, sizeof(buf), "/usr/share/factory", path);
+				arg = buf;
+			}
+			if (!strc) {
+				struct dirent **namelist;
+				int num;
+
+				if (!S_ISDIR(st.st_mode))
+					break;
+				num = scandir(path, &namelist, NULL, NULL);
+				free(namelist);
+				if (num >= 3)
+					break; /* not empty */
```

Comment:
Here we will (potentially) leak memory.  You need to iterate over all the num, when num > 0, freeing all the entries in namelist[i], also handling the case when num == -1.

I suggest splitting out this code to a helper function, e.g., `is_dir_empty()` or similar.

Reply:
great idea! see additional commit i have pushed